state/prune.go

```diff
@@ -0,0 +1,219 @@
+package state
+
+import (
+	"fmt"
+	"gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+	"time"
+
+	"github.com/juju/errors"
+)
+
+// PruneCollection removes collection entries until
+// only entries newer than <maxLogTime> remain and also ensures
+// that the collection is smaller than <maxLogsMB> after the
+// deletion.
+func pruneCollection(mb modelBackend, maxHistoryTime time.Duration, maxHistoryMB int, collectionName string, ageField string, timeUnit TimeUnit) error {
+	entries, closer := mb.db().GetRawCollection(collectionName)
+	defer closer()
+
+	p := collectionPruner{
+		st:       mb,
+		coll:     entries,
+		maxAge:   maxHistoryTime,
+		maxSize:  maxHistoryMB,
+		ageField: ageField,
+		timeUnit: timeUnit,
+	}
+	if err := p.validate(); err != nil {
+		return errors.Trace(err)
+	}
+	if err := p.pruneByAge(); err != nil {
+		return errors.Trace(err)
+	}
+	return errors.Trace(p.pruneBySize())
+}
+
+const historyPruneBatchSize = 1000
+const historyPruneProgressSeconds = 15
+
+type doneCheck func() (bool, error)
+
+type TimeUnit string
+
+const (
+	NanoSeconds TimeUnit = "nanoseconds"
+	GoTime      TimeUnit = "goTime"
+)
+
+type collectionPruner struct {
+	st   modelBackend
+	coll *mgo.Collection
+
+	maxAge  time.Duration
+	maxSize int
+
+	ageField string
+	timeUnit TimeUnit
+}
+
+func (p *collectionPruner) validate() error {
+	if p.maxSize < 0 {
+		return errors.NotValidf("non-positive max size")
+	}
+	if p.maxAge < 0 {
+		return errors.NotValidf("non-positive max age")
+	}
+	if p.maxSize == 0 && p.maxAge == 0 {
+		return errors.NotValidf("backlog size and age constraints are both 0")
+	}
+	return nil
+}
+
+func (p *collectionPruner) pruneByAge() error {
+	if p.maxAge == 0 {
+		return nil
+	}
+
+	t := p.st.clock().Now().Add(-p.maxAge)
+	var age interface{}
+
+	if p.timeUnit == NanoSeconds {
+		age = t.UnixNano()
+	} else {
+		age = t
+	}
+
+	iter := p.coll.Find(bson.D{
+		{"model-uuid", p.st.modelUUID()},
+		{p.ageField, bson.M{"$lt": age}},
+	}).Select(bson.M{"_id": 1}).Iter()
+
+	modelName, err := p.st.modelName()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	logTemplate := fmt.Sprintf("%s age pruning (%s): %%d rows deleted", p.coll.Name, modelName)
```

Comment:
The doubled % in `%%d` means "emit a literal %" - it's not a placeholder. This is a bit tricky because it's a template that produces another template. It substitutes in some values that are known now (collection name and model name) and then produces a template with one slot. (A bit like function currying.)

So it will look like:
```
collection age pruning (model): %d rows deleted
```
... then the current count will be substituted in when reporting progress.